inc/az_ulib_ustream_forward_base.h

```diff
@@ -0,0 +1,374 @@
+      uint8_t* const buffer,
+      size_t buffer_length,
+      size_t* const size);
+
+  /** Concrete `get_remaining_size` implementation. */
+  az_result (*get_remaining_size)(
+        az_ulib_ustream_forward* ustream_forward_instance, 
+        size_t* const size);
+
+  /** Concrete `dispose` implementation. */
+  az_result (*dispose)(
+        az_ulib_ustream_forward* ustream_forward_instance);
+} az_ulib_ustream_forward_interface;
+
+/**
+ * @brief   Signature of the function to release memory passed to the ustream_forward
+ *
+ * @param[in]   release_pointer       void pointer to memory that needs to be free'd
+ *
+ */
+typedef void (*az_ulib_release_callback)(void* release_pointer);
+
+/**
+ * @brief   Pointer to the data from which to read
+ *
+ * void pointer to memory where the data is located or any needed controls to access the data.
+ * The content of the memory to which this points is up to the ustream_forward implementation.
+ *
+ */
+typedef void* az_ulib_ustream_forward_data;
+
+/**
+ * @brief   Structure for data control block
+ *
+ * For any given ustream_forward that is created, one control block is created and initialized.
+ *
+ * @note    This structure should be viewed and used as internal to the implementation of the
+ *          ustream_forward. Users should therefore not act on it directly and only allocate the memory
+ *          necessary for it to be passed to the ustream_forward.
+ *
+ */
+typedef struct az_ulib_ustream_forward_data_cb_tag
+{
+  /** The #az_ulib_ustream_forward_interface* for this ustream_forward instance type. */
+  const az_ulib_ustream_forward_interface* api;
+
+  /** The #az_ulib_ustream_forward_data* pointing to the data to read. It can be anything that a given
+   * ustream_forward implementation needs to access the data, whether it be a memory address to a buffer,
+   * another struct with more controls, etc */
+  const az_ulib_ustream_forward_data* ptr;
+
+  /** The `volatile long` with the number of references taken for this memory. */
+  volatile long ref_count;
+
+  /** The #az_ulib_release_callback to call to release `ptr` once the `ref_count` goes to zero. */
+  az_ulib_release_callback data_release;
+
+  /** The #az_ulib_release_callback to call to release the #az_ulib_ustream_forward_data_cb once the
+   * `ref_count` goes to zero */
+  az_ulib_release_callback control_block_release;
+
+} az_ulib_ustream_forward_data_cb;
+
+/**
+ * @brief   Structure for instance control block
+ *
+ * A reference to the #az_ulib)ustream_forward_data_cb* control_block with member variables to keep
+ *    track of the read offset while iterating over the ustream_forward `Data Source`.
+ *
+ * @note    This structure should be viewed and used as internal to the implementation of the
+ *          ustream_forward. Users should therefore not act on it directly and only allocate the memory
+ *          necessary for it to be passed to the ustream_forward.
+ */
+struct az_ulib_ustream_forward_tag
```

Comment:
>az_ulib_ustream_forward_tag [](http://example.com/codeflow?start=7&length=27)

Because there is no clone, you don't need this struct, only add `length` and `current_position` in the `az_ulib_ustream_forward_data_cb`